src/code/audio/code_8019AF00.c

```diff
@@ -2047,80 +3060,388 @@ void AudioOcarina_TerminaWallGenerateNotes(void) {
-    sOcarinaSongNotes[OCARINA_SONG_MEMORYGAME][sOcarinaMemoryGameAppendPos + 1].length = 0;
+    sOcarinaSongNotes[OCARINA_SONG_MEMORY_GAME][sOcarinaMemoryGameAppendPos].pitch = OCARINA_PITCH_NONE;
+    sOcarinaSongNotes[OCARINA_SONG_MEMORY_GAME][sOcarinaMemoryGameAppendPos].length = 0;
+    sOcarinaSongNotes[OCARINA_SONG_MEMORY_GAME][sOcarinaMemoryGameAppendPos + 1].pitch = OCARINA_PITCH_NONE;
+    sOcarinaSongNotes[OCARINA_SONG_MEMORY_GAME][sOcarinaMemoryGameAppendPos + 1].length = 0;
     if (1) {}
     return false;
 }
-#undef OCARINA_SONG_MEMORYGAME
+#undef OCARINA_SONG_MEMORY_GAME
+
+void AudioOcarina_Update(void) {
+    sOcarinaUpdateTaskStart = gAudioContext.totalTaskCount;
+    if (sOcarinaInstrumentId != OCARINA_INSTRUMENT_OFF) {
+        if (sIsOcarinaInputEnabled == true) {
+            AudioOcarina_ReadControllerInput();
+        }
+
+        if ((sPlaybackState == 0) && (sIsOcarinaInputEnabled == true)) {
+            AudioOcarina_PlayControllerInput(false);
+        }
+
+        if (sOcarinaFlags != 0) {
+            if (sOcarinaFlags & 0x40000000) {
+                AudioOcarina_CheckSongsWithoutMusicStaff();
+            } else {
+                AudioOcarina_CheckSongsWithMusicStaff();
+            }
+        }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/code_8019AF00/func_8019D600.s")
+        AudioOcarina_PlaybackSong();
+        sOcarinaPlaybackTaskStart = sOcarinaUpdateTaskStart;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/code_8019AF00/func_8019D758.s")
+        if (sPlaybackState == 0) {
+            AudioOcarina_RecordSong();
+        }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/code_8019AF00/func_8019D864.s")
+        if ((sOcarinaFlags != 0) && (sPrevOcarinaPitch != sCurOcarinaPitch)) {
+            sOcarinaDropInputTimer = 1; // Drops ocarina input for 1 frame
+        }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/code_8019AF00/func_8019D8B4.s")
+        sPrevOcarinaPitch = sCurOcarinaPitch;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/code_8019AF00/func_8019D8E4.s")
+        if (sOcarinaDisableTimer != 0) {
+            sOcarinaDisableTimer--;
+            if (sOcarinaDisableTimer == 0) {
+                AudioOcarina_SetInstrument(OCARINA_INSTRUMENT_OFF);
+            }
+        }
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/code_8019AF00/func_8019DF28.s")
+    AudioOcarina_UpdatePlayingStaff();
+    AudioOcarina_UpdatePlaybackStaff();
+    AudioOcarina_UpdateRecordingStaff();
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/code_8019AF00/func_8019DF64.s")
+void AudioOcarina_PlayLongScarecrowAfterCredits(void) {
+    static u8 sScarecrowAfterCreditsState = 0;
+    static u8 sScarecrowAfterCreditsIntrumentId = OCARINA_INSTRUMENT_DEFAULT;
+    static u16 sScarecrowAfterCreditsTimer = 1200;
+
+    switch (sScarecrowAfterCreditsState) {
+        case 0:
+            if (sScarecrowAfterCreditsTimer-- == 0) {
+                // OoT originally had 7 Ocarina instruments
```

Comment:
What's the difference between what normal people think of as instruments and the number you're talking about here?

Reply:
Not having read this whole PR yet, There is an array of Audiobank 00 instrument indexes of roughly size 16 that contained instruments that Link could use as instruments when playing the Ocarina (Ocarina for Human Link, Trumpet for Deku Link, Drum for Darmani, Guitar for Mikau) but it also contains instruments that other characters can use, mostly when teaching link new songs (The monkey sound when teaching Sonata, Tatl's bell she uses to teach link Oath to order, King Ikana's "Whoa!" sfx when teaching Elegy, Japas' Bass Guitar for the jam session, ect.)

At least I think that's what this comment is referring to, maybe that list doesn't need as many instruments in it in OoT (Shieks harp and Impa's whistle come to mind, but there are no forms in that game so that shaves off at least 3)

Reply:
isghj5 is exactly right with these ocarina instruments.

This loop should really be looping through all ocarina instruments, `< OCARINA_INSTRUMENT_MAX`. This value was 7 in ocarina of time, but jumps up to 17 in MM. Obviously this was just copied over from OoT, so while the `7` was the max there, it's just an arbitrary enum in MM, so saying `< OCARINA_INSTRUMENT_GORON_DRUMS` doesn't really make sense. So hence the comment

Reply:
Mm, maybe it should have an actual define?

Reply:
You mean something like `OCARINA_INSTRUMENT_OOT_MAX`?

Reply:
Yeah, wouldn't be the first time they didn't convert something properly when copypasting, as I'm sure you can sympathise with.